src/input/return_enums.rs

```diff
@@ -77,33 +131,95 @@ fn validate_iter_to_vec<'a, 's>(
     }
 }
 
-macro_rules! any_next_error {
-    ($py:expr, $err:ident, $input:ident, $index:ident) => {
-        ValError::new_with_loc(
-            ErrorType::IterationError {
-                error: py_err_string($py, $err),
-            },
-            $input,
-            $index,
-        )
-    };
+pub trait BuildSet {
+    fn build_add(&self, item: PyObject) -> PyResult<()>;
+
+    fn build_len(&self) -> usize;
 }
 
-macro_rules! generator_too_long {
-    ($input:ident, $index:ident, $max_length:expr, $field_type:ident) => {
-        if let Some(max_length) = $max_length {
-            if $index > max_length {
-                return Err(ValError::new(
-                    ErrorType::TooLong {
-                        field_type: $field_type.to_string(),
-                        max_length,
-                        actual_length: $index,
-                    },
-                    $input,
-                ));
+impl BuildSet for &PySet {
+    fn build_add(&self, item: PyObject) -> PyResult<()> {
+        self.add(item)
+    }
+
+    fn build_len(&self) -> usize {
+        self.len()
+    }
+}
+
+impl BuildSet for &PyFrozenSet {
+    fn build_add(&self, item: PyObject) -> PyResult<()> {
+        unsafe {
+            py_error_on_minusone(
+                self.py(),
+                ffi::PySet_Add(self.as_ptr(), item.to_object(self.py()).as_ptr()),
+            )
+        }
+    }
+
+    fn build_len(&self) -> usize {
+        self.len()
+    }
+}
+
+#[allow(clippy::too_many_arguments)]
+fn validate_iter_to_set<'a, 's>(
+    py: Python<'a>,
+    set: impl BuildSet,
+    iter: impl Iterator<Item = PyResult<&'a (impl Input<'a> + 'a)>>,
+    input: &'a (impl Input<'a> + 'a),
+    field_type: &'static str,
+    max_length: Option<usize>,
+    validator: &'s CombinedValidator,
+    extra: &Extra,
+    definitions: &'a [CombinedValidator],
+    recursion_guard: &'s mut RecursionGuard,
+) -> ValResult<'a, ()> {
+    let mut errors: Vec<ValLineError> = Vec::new();
+    for (index, item_result) in iter.enumerate() {
+        let item = item_result.map_err(|e| any_next_error!(py, e, input, index))?;
+        match validator.validate(py, item, extra, definitions, recursion_guard) {
+            Ok(item) => {
+                set.build_add(item)?;
+                if let Some(max_length) = max_length {
+                    let actual_length = set.build_len();
```

Comment:
I think with some smarter logic we could remove a bunch of the calls to len(set) here, not sure it's worth it though.

What I'm thinking is, you have a counter that tells you how many distinct items you can add before exceeding the limit, and you track that in rust, subtracting one from it every time you hit something. When _that_ counter goes negative, only then you check if the set is too long. If it isn't, you reset the counter telling you how many more distinct items you can add.

I don't think the implementation would actually require many lines of code, and it seems to me a nontrivial amount of overhead e.g. if creating a set of 1000 items with max len 2000. But I don't know, maybe it is negligible, and either way maybe not worth it.

Reply:
Very good idea.

Reply:
done.

Reply:
I've tried that, but it's marginally degraded performance - probably an anomaly, but I doesn't seem to have improved performance as we hoped.

Please review my implementation, and check if I've done something dumb.

Reply:
okay, I've added a dedicated benchmark for this, and it seems this approach doesn't improve performance, even in the optimal cases where in the input has no duplicates.

I guess the check to set length is very fast.